vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/dataview/GridListDataViewTest.java

```diff
@@ -0,0 +1,119 @@
+    public void dataViewForFaultyDataProvider_throwsException() {
+        DataProvider<String, Void> dataProvider = DataProvider
+                .fromCallbacks(query -> Arrays.asList("one").stream(),
+                        query -> 1);
+        exceptionRule.expect(IllegalStateException.class);
+        exceptionRule.expectMessage(
+                "GridListDataView only supports 'ListDataProvider' or it's subclasses, but was given a 'AbstractBackEndDataProvider'");
+
+        Grid<String> grid = new Grid<>();
+        grid.setDataProvider(dataProvider);
+        new GridListDataView<String>(grid);
+    }
+
+    @Test
+    public void dataViewWithItems_returnsExpectedItemsForMethods() {
+        String[] items = new String[] { "item1", "item2", "item3", "item4" };
+        Grid<String> grid = new Grid<>();
+        grid.setDataProvider(DataProvider.ofItems(items));
+        GridListDataView<String> dataView = new GridListDataView<>(grid);
+
+        // Test getItemOnRow returns correct item
+        Assert.assertEquals("Wrong item returned for row", items[2],
+                dataView.getItemOnRow(2));
+
+        // Test hasNext-/-PreviousItem
+        Assert.assertTrue("Item in middle should have next item",
+                dataView.hasNextItem(items[2]));
+        Assert.assertTrue("Item in middle should have previous item",
+                dataView.hasPreviousItem(items[1]));
+
+        Assert.assertTrue("First item should have next item",
+                dataView.hasNextItem(items[0]));
+        Assert.assertTrue("Last item should have previous item",
+                dataView.hasPreviousItem(items[3]));
+
+        Assert.assertFalse("No next item for last item should be available",
+                dataView.hasNextItem(items[3]));
+        Assert.assertFalse(
+                "No previous item for first item should be available",
+                dataView.hasPreviousItem(items[0]));
+
+        // Test getNext-/-PreviousItem
+        Assert.assertEquals("Faulty next item", items[3],
+                dataView.getNextItem(items[2]));
+        Assert.assertEquals("Faulty previous item", items[1],
+                dataView.getPreviousItem(items[2]));
+
+        Assert.assertNull("Got next item for last item",
+                dataView.getNextItem(items[3]));
+        Assert.assertNull("Got previous item for first index",
+                dataView.getPreviousItem(items[0]));
+
+        // Test getSize
+        Assert.assertEquals("Unexpected size for data", items.length,
+                dataView.getDataSize());
+
+        // Test containsItem
+        Assert.assertTrue("Set item was not found in the data",
+                dataView.isItemPresent(items[3]));
+        Assert.assertFalse("Non existent item found in data",
+                dataView.isItemPresent("item6"));
+    }
+
+    @Test
+    public void dataView_withFilter_methodsUseFilteredData() {
+        String[] items = new String[] { "item1", "item2", "item3", "item4" };
```

Comment:
non-blocking: withSortComparator test is missing. However, it could be done within Filtering/Sorting ticket #8313